src/plugins/data-category-tweaks/index.tsx

```diff
@@ -0,0 +1,285 @@
+    ScratchBlocks.Flyout.prototype.show = function (xmlList) {
+      this.workspace_.registerToolboxCategoryCallback("VARIABLE", variableCategoryCallback);
+      this.workspace_.registerToolboxCategoryCallback("LIST", listCategoryCallback);
+      return oldShow.call(this, xmlList);
+    };
+
+    // Use Scratch's extension category mechanism to replace the data category with our own.
+    // https://github.com/scratchfoundation/scratch-gui/blob/ddd2fa06f2afa140a46ec03be91796ded861e65c/src/containers/blocks.jsx#L344
+    // https://github.com/scratchfoundation/scratch-gui/blob/2ceab00370ad7bd8ecdf5c490e70fd02152b3e2a/src/lib/make-toolbox-xml.js#L763
+    // https://github.com/scratchfoundation/scratch-vm/blob/a0c11d6d8664a4f2d55632e70630d09ec6e9ae28/src/engine/runtime.js#L1381
+    const originalGetBlocksXML = runtime.getBlocksXML;
+    runtime.getBlocksXML = function (target) {
+      const result = originalGetBlocksXML.call(this, target);
+
+      if (hasSeparateListCategory) {
+        result.push({
+          id: "data",
+          xml: `
+          <category
+            name="%{BKY_CATEGORY_VARIABLES}"
+            id="variables"
+            colour="${ScratchBlocks.Colours.data.primary}"
+            secondaryColour="${ScratchBlocks.Colours.data.tertiary}"
+            custom="VARIABLE">
+          </category>
+          <category
+            name="${msg("plugins.dataCategoryTweaks.list")}"
+            id="lists"
+            colour="${ScratchBlocks.Colours.data_lists.primary}"
+            secondaryColour="${ScratchBlocks.Colours.data_lists.tertiary}"
+            custom="LIST">
+          </category>`,
+        });
+        result.map = (callback) => {
+          // Prevent Scratch from trying to change the color of the added category in high contrast mode.
+          // https://github.com/scratchfoundation/scratch-gui/blob/44eb578/src/containers/blocks.jsx#L358-L361
+          // https://github.com/scratchfoundation/scratch-gui/blob/44eb578/src/lib/themes/blockHelpers.js#L18-L53
+          return Array.prototype.map.call(result, (extension) => {
+            if (extension.id === "data") return extension;
+            else return callback(extension);
+          });
+        };
+      }
+      return result;
+    };
+
+    // If editingTarget is set, the editor has already rendered and we have to tell it to rerender.
+    if (vm.editingTarget) {
+      (vm as any).emitWorkspaceUpdate();
+    }
+
+    const dynamicEnableOrDisable = () => {
+      if (vm.editingTarget && hasSeparateListCategory) {
+        (vm as any).emitWorkspaceUpdate();
+      }
+
+      if (workspace && hasSeparateLocalVariables && hasMoveReportersDown) {
+        workspace.refreshToolboxSelection_();
+      }
+        
+      hasSeparateListCategory = false
+      hasSeparateLocalVariables = false
+      hasMoveReportersDown =  false
+    }
+
+  return () => {
+    dynamicEnableOrDisable()
+    register.dispose()
```

Comment:
You need to restore the previously overridden events here, otherwise, after the plugin is remounted, flags like `hasSeparateListCategory` won't get the latest data and will always be `false`.
 > `runtime.getBlocksXML = originalGetBlocksXML;ScratchBlocks.Flyout.prototype.show = oldShow;`